pilot/pkg/bootstrap/configcontroller.go

```diff
@@ -268,6 +276,41 @@ func mcpSecurityOptions(ctx context.Context, configSource *meshconfig.ConfigSour
 	return securityOption, nil
 }
 
+// initInprocessAnalysisController spins up an instance of Galley which serves no purpose other than
+// running Analyzers for status updates.  The Status Updater will eventually need to allow input from istiod
+// to support config distribution status as well.
+func (s *Server) initInprocessAnalysisController(args *PilotArgs) error {
+
+	processingArgs := settings.DefaultArgs()
+	processingArgs.KubeConfig = args.Config.KubeConfig
+	processingArgs.EnableValidationController = false
+	processingArgs.EnableValidationServer = false
+	processingArgs.MonitoringPort = 0
+	processingArgs.Liveness.UpdateInterval = 0
+	processingArgs.Readiness.UpdateInterval = 0
+	processingArgs.Insecure = true // TODO - use sidecar?
+	processingArgs.EnableServer = false
+	processingArgs.MeshConfigFile = args.Mesh.ConfigFile
+	processingArgs.EnableConfigAnalysis = true
+
+	processing := components.NewProcessing(processingArgs)
+
+	s.addStartFunc(func(stop <-chan struct{}) error {
```

Comment:
Only one controller instance (pod) should be updated CRD status at a time. You can use s.leaderElection.AddRunFunction instead which will integrate with the existing leader election logic.

Reply:
Are we starting a full on gRPC server here?

Reply:
@howardjohn no, we are starting a status controller.

@ayj my understanding is that the galley status controller already solves for the leader election problem by debouncing status updates and only writing updates when there is actually a change.  This was functional in 1.4, with galley scaled horizontally, so I don't think anything should need to be changed to support horizontal scale in istiod.

Reply:
NewProcessing() will start a gRPC server which isn't needed to enable status updates. You could update processing.Start() to not start the server if EnableServer is false.

As istiod scales we don't need each replica watching and updating status in parallel.

Reply:
As we add config distribution status to istiod in 1.6, every istiod instance will be required to participate in status updates, as each istiod will have distribution information for a fraction of the envoy proxies.  I don't see any way around this...

With leader election, how do you handle the lead istiod being shut down?  Is there a become_leader event?

Reply:
leader election is already fully handled by the library and handles shutdown, etc.

if I have 3 istiods, which have different views of distribution, all 3 are going to write to the same CR? How does that work?

Reply:
Each of them writes only a part of the status - the part they are familiar with.  Because the envoys are divided among the three istiod's, they each have a part of the picture.  Each one writes the number of envoys which have acked the changed and the number of envoys which have not.  These numbers are added up using a pessimistic concurrency model to come up with the total config distribution status.

Reply:
Can you link an design about what they are writing? I can't wrap my head around how this can work (yet)

Reply:
I am digging through history to find the doc.  Since this particular doc is for config distribution status, which this PR does not achieve, can we continue this part of the conversation in a different channel?

Reply:
Sure.

I don't status controller should have multiple replicas though. We had issues with split brain in the past where they fight each other and get into a reconcile loop. From other projects I have seen leader election always used for status. Are there any reasons why we should not do it here?

Reply:
I can implement it here, if that is what it takes to get this PR merged, but it will need to be undone in the next week to facilitate Config Distribution Status.  I am generally not in favor of implementing throw away code, but it does look like it's only a few lines...

When you reference past split brain issues, are you referring to Status Controllers?  I am not as familiar with implementation as I'd like, but Oz and Martin designed the status controller to specifically avoid such a scenario... Unfortunately, we no longer have them to weigh in...

Reply:
Yes. We have had a status controller in pilot for a very long time.

If we are going to make decisions based on  Config Distribution Status then that is in scope for this PR and I would like to see the design before deciding anything.

Reply:
Also my gRPC concerns were not addressed I believe, just want to make sure they do not get lost in the thread here

Reply:
Re: Status Controller.  I understand now.  The Status Controller used here is not Pilot's Status Controller (istio.io/pilot/pkg/config/kube/ingress/StatusUpdater), but Galley's (istio.io/galley/pkg/config/source/kube/apiserver/status/Controller), which has built in protection against dual brain.  I will see if I can find an old design doc from Oz or Martin describing how that works.

Reply:
My most recent commit is intended to address the GRPC concern.

Reply:
We only need one instance to perform CRD analysis. Leader election is a suitable solution to that problem and we already have support.  Use `s.leaderElection.AddRunFunction` instead of `s.addStartFunc`.

For config status distribution, in addition to any split brain issues, it also puts additional load on the kube-apiserver. This could be particularly bad in larger clusters, e.g. lots of config x lots of istiod instances.

Reply:
Based on video chat with Jason and John, moving analysis to istiod's leader election so as only to run one.  Will need different solution for distribution status.